New pattern proposal: InnerSource Hackathon (#700)
gruetter/innerSourceTest: patterns/1-initial/innersource-hackathon.md

## Title

> InnerSource Hackathon

Reviewer: Just brainstorming:
If you could not use "InnerSource" in the title, how might you call this pattern?

Author: I am not sure. The closest I can think off is - 'Hackathon: Safe experiment space', but I feel it doesn't sound right. I took inspiration from other patterns - InnerSource Portal and InnerSource License, which give the user a good understanding of what to expect in the pattern, just by the title. 

Please feel free to suggest one :)

Reviewer: Naming things is hard, as we know ;)

I don't have a better idea right now idea.

I asked about this because all of these are InnerSource patterns, so naturally the word InnerSource will appear in a lot of places. Therefore we try to not overuse the word in the pattern titles, to distinguish the patterns from each other a bit more.

For now:
Let's just keep the title as is.
Once the patterns matures more through feedback from other organizations, maybe we have an idea for a different title.

Reviewer 2: Maybe it would help to include something that the Hackathon achieves in the title?

Something that expresses the gained practical experience?

Though I can follow the argument of @spriya-m referring to InnerSource Portal and -License as examples.

## Patlet

> In a company, initially only InnerSource enthusiasts are interested and practicing InnerSource during the early stages of InnerSource adoption; not all engineering teams are willing or have enough time and resources to adopt InnerSource. In this scenario, it is good to provide a safe space to try and adopt InnerSource through an InnerSource Hackathon event within the company.

Reviewer 2: I love this pattern because it ties into the [crossing the chasm](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/1-initial/crossing-chasm.md) pattern: It's a lovely way to get people to practice instead of only learn or hear about best practices.

## Problem

> The company wants to adopt InnerSource as software development methodology but only those familiar with open source principles or those who understand the benefits of InnerSource, adopt it. It results in just a handful of InnerSource projects and is difficult to scale beyond that.

Reviewer 2: Would you do that purely as a Hackthon targeted at learning about InnerSource best practices? 

I'm asking because reading your suggested pattern, made me remember some of the very early InnerSource days at my employer: There was a yearly hackathon established already. Participating teams decided to try out InnerSource best practices there early on - come to think of it, that was one of the occasions where best practices spread across team boundaries.

Author: Yes, the goal is to make the engineering community learn about InnerSource best practices. When ISPO or sometimes OSPO (driving InnerSource as well) is not able to reach the entire engineering community, then such an event would help draw some attention. 

Just out of curiosity, the yearly hackathon that you mentioned - was it specific to InnerSource? If so, can we quote it as a known instance?

Reviewer 2: The one I mentioned wasn't specific to InnerSource - it had been established long before the InnerSource initiative.

What happened was that some Hackathon teams would naturally adopt InnerSource for those few days to try it out. As a result the practice was adopted by new teams afterwards.

I think this shows just how strong a Hackathon as a pattern is: Even if it is not focused 100% on spreading the practice of InnerSource it still helps teams adopt those new ways of working. We could extend the pattern such that there are two Hackathon options - those organised by ISPO/OSPO and 100% geared towards InnerSource exercises plus more generic Hackathons than we can of course include my employer as well.

## Context

> Scenario 1:

Reviewer: If you were to give these scenarios names, how would you call them?
What I am after is the essence of each scenario.

Author: I would say - 
Scenario 1: Challenges in scaling beyond the early adopters
Scenario 2: Challenges in getting contributions and building community around InnerSource projects

Reviewer: I have worked these proposals into subsection headings. 👍

* not familiar with InnerSource and being ignorant to know about it
* not enough time to prioritize InnerSource, given the regular work deliverables
* relunctance to changing ways of working when everything works well already
* perception that InnerSource requires more work and responsibilities

Scenario 2:
Teams slowly start adopting InnerSource and open up their repositories. Some new projects are started as InnerSource projects from the start (that is open for contributions). However, there are not many contributors. It is also challenging to build a community around those projects, in spite of publishing the projects in an InnerSource Portal.  This could be due to various reasons like:

* engineers do not have time to explore new InnerSource projects and contribute, outside regular work deliverables
* no additional incentive to this effort apart from being acknowledged
* lack of motivation by middle management

## Forces

Apart from the reasons listed in the Context section, there are other factors from different entities that prevent teams from adopting InnerSource like:

> * From middle management: perception of no direct benefit to the team apart from team upskilling and individual growth; eventually resulting in no motivation by the middle management to the engineers in trying out InnerSource ways of working.

Reviewer 2: @spier we already have some patterns addressing push back like that - do you think it would make sense to link to those from here?

E.g.: [Change middle management mindset](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/1-initial/change-the-middle-management-mindset.md), [Developer incentive alignment](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/1-initial/developer-incentive-alignment-for-innersource-contribution.md), [Incentivise voluntary contributions](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/1-initial/incentive-mechanisms-for-voluntary-contribution.md), [Overcome time constraints](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/1-initial/overcoming-project-management-time-pressures.md)?

Reviewer: Nice thought! We certainly could link to some of these, yes.

Just note that those other patterns have not reached the "Structured" [maturity level](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/meta/contributor-handbook.md#in-a-nutshell) yet. i.e. they may have not bee confirmed by an organization that used these patterns.

Reviewer 2: Oh, right. Is there another good place to track these links?

Reviewer: Have no good yet, no. 

Will just leave this conversation open, while already going ahead and merging the PR.

* From engineers: investing time to try new ways of working when there is already time and resource constraint for regular work deliverables; it is also challenging to balance both and show outcomes in both; this curbs motivation from the engineers as there is little safe space to experiment.
* From senior leadership: not prioritizing investment in InnerSource ways of working such as it being part of long term goals or OKRs.

For those new to InnerSource or any technology/methodology, there is a need for safe space to try and experiment without definite outcomes in the initial stages. These factors and forces hinder such an experiment ground.

## Solution

> Organize a company wide hackathon focused on InnerSource. An InnerSource hackathon provides a safe space for engineers to try and contribute to InnerSource projects without any presumption and prejudice. It could be a 1 or 2 day event.

Reviewer 2: Thinking of different organisational sizes - do you think there is a sweet spot where company wide makes sense vs. at which point it gets too large to be effective?

Author: I agree to your point - maybe it can be conducted as a business unit level in large organizations?

Reviewer: Also depends on what you want to achieve with InnerSource in the first place.

e.g. if the purpose is to encourage re-use across department borders in a large multi-national org, then you would likely want the hackathon to span all those departments. Even if that means that the hackathon becomes larger.

Different hackathon formats and sizes have different pros and cons. But I think all of them can be effective. Sometimes you have to "encourage then good" e.g. for the prior example of the multi-national org, if you want cross-team collaboration to happen, then you can setup soft rules like "only teams that contain members from at least 2 different teams from 2 separate engineering offices are considered when choosing an overall winner".

> Organize a company wide hackathon focused on InnerSource. An InnerSource hackathon provides a safe space for engineers to try and contribute to InnerSource projects without any presumption and prejudice. It could be a 1 or 2 day event.

Reviewer: ```suggestion
Organize a company-wide hackathon focused on InnerSource. An InnerSource hackathon provides a safe space for engineers to try and contribute to InnerSource projects without any presumption and prejudice. It could be a 1 or 2 day event.
```

> It can preferrable be organized by InnerSource Program Office (ISPO), if it exists in the organization or by the Open Source Program Office (OSPO), if OSPO also drives InnerSource within the company. It can also be organized by other common service organizations too, if need be. Basically a central team or a group of individuals, who believe in InnerSource, can organize the hackathon.

Reviewer 2: Would piggy-backing on an existing hackathon work as well?

Author: IMO, piggy-backing with another hackathon would dilute the focus on InnerSource. Having said that, it depends on each company too. For a company in early stages of InnerSource with not many InnerSource practitioners, it is better to have it focused just on InnerSource.

Reviewer 2: Now that I read this: In my case, where piggy-packing did happen, it occured at a later stage where several engineers already had experience. Essentially we started with the "get started as an experiment", scaled in a sub org after that and crossed towards other departments through the hackathon.

So maybe the piggy-backed version of this pattern would then work better for companies that already have several practitioners to carry adoption further. Maybe also to carry learnings concerning the practice between teams.

All engineers in the organization can participate in the hackathon. The participants can be new to InnerSource, or InnerSource practitioners already. They can participate individually or as a team. Participating as a team also provides a safe environment, for example, those are new to InnerSource can team up InnerSource practitioners.

There could be one or more categories to participate as follows:

* Start a new InnerSource project: It could be either starting a new project from scratch as InnerSource or making a existing project InnerSource ready.
* Contribute to the existing InnerSource project: InnerSource project owners can list down the features in GitHub issues as a pre-requisite before the hackathon and the participants can contribute to it during the hackathon.

> * Participate as an InnerSource mentor: Participants can also mentor new InnerSource practitioners and provide knowledge transfer or coaching during the event.

Reviewer: ❤️ love this section here!

The event could be held virtually or physically in different locations in the organization based on the cost and budget involved. A virtual event will have best ROI but it could depend on each organization.

> The winners and participants should be recognized and acknowledged in a company wide forum at the end of the hackathon. This is important as it keeps motivating them and more engineers to adopt and practice InnerSource going forward.

Reviewer: ```suggestion
The winners and participants should be recognized and acknowledged in a company-wide forum at the end of the hackathon. This is important as it keeps motivating them and more engineers to adopt and practice InnerSource going forward.
```

Reviewer: I have organized a number of hackathons in the past, and for company-internal hackathons the part of "winning" was always controversial. Some people loved the friendly competition, others thought that selecting a winner was harmful.

Personally I like the friendly competition :)

For an InnerSource hackathon in particular I am curious:
How did you select the winners? Which criteria did you apply?

Author: Yes, it is tough to set the right judging criteria :) 

We had two categories in the hackathon - 'Start a new InnerSource project' and 'Contribute to existing InnerSource projects'. In the contributing category, as a pre-requisite, the innersource maintainers listed down the GitHub issues for which they were open for contributions. The participants can contribute to one of these issues or make their own contribution.

Judging criteria for 'Start a new InnerSource project': 
- How well the InnerSource project aligns with our internal guidelines for an InnerSource project (InnerSource readiness)? 
- What is the value proposition? How does it benefit the many in the company? 
- How welcoming and open the repository is for new contributors? 

Judging criteria for 'Contribute to an existing InnerSource project': 
- Participating team should contribute to an InnerSource project that they don't own. 
- Value-add the contribution brings to the project. 
- How well it aligns with the project?
- How well the contributor has aligned with the contributing guidelines in the project? 

The judging criteria was shared with the judging panel as well as the participants prior to the event. The winners were selected based on these criteria. We also declared winners per category as it would not be fair to compare outcome of different categories.

Such an event provides a safe space for the engineers who want to adopt or contribute to InnerSource but didn't have the time and motivation to do it or for those who kept deprioritizing due to high priority work deliverables. From middle management point of view, 1 or 2 days for such an event is not much of an ask and hence they are more likely to accept.

## Resulting Context

* There is increased adoption of InnerSource by providing such a safe platform to try and experiment combined with recognition and support by the senior leadership and middle management.

> * It results in more projects being made InnerSource ready and innersourced, and also results in more InnerSource contributions and start building community around existing InnerSource projects.

Reviewer: The phrase "being made InnerSource ready and innersourced" caught my eye.

How do you differentiate between "made InnerSource ready" and "innersourced"?

Author: A project can be InnerSource ready if it has minimum pre-requisites met. For example:
- if a repository has [standard base documentation](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/2-structured/base-documentation.md) (and/or)
- if the corresponding project maintainers have agreed to innersource the project and set aside time for addressing queries or contributions from the community (and/or) 
- if there is a proper [communication tooling](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/2-structured/communication-tooling.md) for the project, that the community can use to reach the maintainers easily. 

By innersourced, I mean that the project is announced as an InnerSource project within the company and officially made open for contributions. It can also include listing the project in the [InnerSource Portal](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/patterns/2-structured/innersource-portal.md) if used by the company. 

If a project is innersourced, it means that the repository and the team have already met InnerSource-ready criteria.

However, these criteria and definitions can differ in each company, but this is what I meant by 'make InnerSource ready' and 'innersourced' :)

Reviewer: Got it. Typically we try not to use the verb form of "innersourced".

Let me try to make a proposal, that would also split up this line into more distinct benefits.

```suggestion
* More InnerSource projects are published within the company.
* The InnerSource projects receive more contributions.
* Communities start to form around these InnerSource projects.
```

* This happens not only during the event but continues after the event too, with the hackathon participants acting as InnerSource ambassadors in their teams.
* It also helps ISPO and OSPO spread awareness about InnerSource best practices quickly, across the whole engineering community in the organization.

> * Such an event also helps discover some cool hobby projects that were developed to solve the needs of a specific team but turns out that many teams have similar requirements.

Reviewer: I think I know what you mean by "hobby projects". However as these happen at work, and are hence not for leisure purposes, I wonder if we can find a different term?

Maybe rephrasing it like this could work?

```suggestion
* Such an event also gives more exposure to projects that were developed to solve the needs of a specific team but turns out that many teams have similar requirements.
```

Reviewer 2: I agree that "hobby" can be misleading. I like the re-phrase of @spier 

However the term "hobby" did make me think of a type of project that typically starts at or around Hackathons that I don't think of reading the rephrase.

What if we added:

* Such an event also helps find time to work on side projects that otherwise are dropped for lack of time or priority. Often those concern issues that all teams have to deal with. Impact locally though often is too low to address them - value only gets visible when looking at the bigger picture.

Think of the once famous Google 20% time that helped trigger GMail as legend has it. I'm trying to capture that as motivation of the Hackathon.

## Title

InnerSource Hackathon

## Patlet

In a company, initially only InnerSource enthusiasts are interested and practicing InnerSource during the early stages of InnerSource adoption; not all engineering teams are willing or have enough time and resources to adopt InnerSource. In this scenario, it is good to provide a safe space to try and adopt InnerSource through an InnerSource Hackathon event within the company.

## Problem

The company wants to adopt InnerSource as software development methodology but only those familiar with open source principles or those who understand the benefits of InnerSource, adopt it. It results in just a handful of InnerSource projects and is difficult to scale beyond that.

## Context

> Scenario 1:

Reviewer: ```suggestion
### Scenario 1: Challenges in scaling beyond the early adopters

```

The senior leadership believes in InnerSource and wants to drive it throughout the company. The engineers who are familiar with open source principles and/or understand the benefits of InnerSource are the early adopters. There is success with these initial pilot project and teams. Now the next step is to drive it across the company. There might be reluctance from engineering teams due to various factors like:

* not familiar with InnerSource and being ignorant to know about it
* not enough time to prioritize InnerSource, given the regular work deliverables
* relunctance to changing ways of working when everything works well already
* perception that InnerSource requires more work and responsibilities

> Scenario 2:

Reviewer: ```suggestion
### Scenario 2: Challenges in getting contributions and building community around InnerSource projects

```
